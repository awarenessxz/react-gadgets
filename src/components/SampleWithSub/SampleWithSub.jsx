import React, { useState } from 'react';
import PropTypes from 'prop-types';

import SubSample from './SubSample/SubSample';

/**
 * Sample Component with Sub Component
 */
const SampleWithSub = props => {
    const [counter, setCounter] = useState(0);
    const [selection, setSelection] = useState('Null');

    const onCounterChange = value => {
        setCounter(counter + value);
    };

    const onSelectClick = () => {
        if (counter < 0 || counter >= props.testingList.length) {
            setSelection('Error!!!!! Index out of bound');
        } else {
            setSelection(props.testingList[counter].name);
        }
    };

    return (
        <div>
            <h1>Select User from List Below</h1>
            <p>There are two parts to this Sample: </p>
            <ol>
                <li>
                    Click on the Add / Minus buttons to change index value. (to demo testing with
                    multiple same child component in parent component)
                </li>
                <li>
                    Select button to select data from a list. (to demo testing data array and error
                    handling)
                </li>
            </ol>
            <h2 data-testid='counter'>Index = {counter}</h2>
            <SubSample onButtonClicks={onCounterChange} />
            <SubSample onButtonClicks={onCounterChange} value={3} />
            <hr />
            <button onClick={onSelectClick}>Select Row based on index above</button>
            <ul>
                {props.testingList.map((item, index) => {
                    return (
                        <li key={index} data-testid='usernames'>
                            {item.name}
                        </li>
                    );
                })}
            </ul>
            <p>Selected User: {selection}</p>
        </div>
    );
};

SampleWithSub.defaultProps = {
    testingList: [
        {
            name: 'Alice',
            email: 'Alice@example.com',
        },
        {
            name: 'Bruce',
            email: 'Bruce@example.com',
        },
        {
            name: 'Charlie',
            email: 'Charlie@example.com',
        },
        {
            name: 'Donnie',
            email: 'Donnie@example.com',
        },
        {
            name: 'Elizabeth',
            email: 'Elizabeth@example.com',
        },
        {
            name: 'Frank',
            email: 'Frank@example.com',
        },
    ],
};

SampleWithSub.propTypes = {
    /** Array of custom data where data example = { name: 'James', email: 'james@example.com' } */
    testingList: PropTypes.array,
};

export default SampleWithSub;
