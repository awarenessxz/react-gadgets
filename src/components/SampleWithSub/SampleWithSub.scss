.wrapper {
  border: 1px dotted cornflowerblue;
  width: 50%;
  margin: 0 auto;
  padding: 5px;
}

.wrapper2 {
  padding: 5px;
}

.wrapper p {
  margin: 0;
  display: inline-block;
}

.wrapper button {
  float: right;
}