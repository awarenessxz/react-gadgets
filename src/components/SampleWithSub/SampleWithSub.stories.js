import React from 'react';

import SampleWithSub from './SampleWithSub';
import SubSample from './SubSample/SubSample';

// telling storybook the component we are documenting
export default {
    title: 'Welcome|SampleV2', // how to refer to the component in the sidebar of the Storybook app,
    component: SampleWithSub, // the component itself
    subcomponents: { SubSample }, // for documenting the child components
};

/************************************************
 * Stories (different permutation of the component)
 ************************************************/

// Note: the first story will be the 'example component' used for DocPage (documentation which are auto generated)
export const SampleWithSubComponent = () => {
    return <SampleWithSub />;
};
