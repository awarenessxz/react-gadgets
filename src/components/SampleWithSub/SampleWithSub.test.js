import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
// Note that: cleanup is called automatically after each test (basically unmounts React trees that were mounted with render)

import SampleWithSub from './SampleWithSub';

// function to render component before each test
const renderComponent = props => {
    return render(<SampleWithSub {...props} />); // returns methods (refer to api)
};

// 1. Testing if component renders properly
describe('Testing if component renders properly', () => {
    // Snapshot testing
    it('Snapshot Testing', () => {
        const { asFragment } = renderComponent();
        expect(asFragment()).toMatchSnapshot();
    });

    // Testing Dom Elements: check if renders with correct value (Note the usage of data-testid in component)
    describe('component renders with correct state (Testing Dom elements)', () => {
        // Test for all the normal HTML DOM elements
        it('all HTML DOM renders correctly', () => {
            renderComponent();

            expect(screen.getByTestId('counter')).toHaveTextContent('Index = 0');
            expect(screen.getByRole('button', { name: 'Add 1' })).toBeTruthy(); // Test Child Component renders
            expect(screen.getByRole('button', { name: 'Add 3' })).toBeTruthy(); // Test Child Component renders
        });

        // Test if array is rendered correctly
        describe('Test if array list is renderer correctly', () => {
            it('renders list', () => {
                const fakeContacts = [
                    { name: 'Georgie', email: 'Georgie' },
                    { name: 'Sheldon', email: 'Sheldon' },
                    { name: 'Missy', email: 'Missy' },
                ];
                renderComponent({ testingList: fakeContacts });

                // get the list by test Id
                const contactNames = screen.getAllByTestId('usernames').map(li => li.textContent);
                const fakeContactNames = fakeContacts.map(c => c.name);
                expect(contactNames).toEqual(fakeContactNames);
                expect(contactNames).toMatchInlineSnapshot(`
                    Array [
                      "Georgie",
                      "Sheldon",
                      "Missy",
                    ]
                `); // alternative
            });
        });
    });
});

// 2. Testing Events (Eg. onClick)
describe('Testing Events', () => {
    describe('Testing Add / Minus Buttons', () => {
        it('Clicking Add 1', () => {
            renderComponent();
            fireEvent.click(screen.getByRole('button', { name: 'Add 1' }));
            expect(screen.getByTestId('counter')).toHaveTextContent('Index = 1');
        });

        it('Clicking Minus 1', () => {
            renderComponent();
            fireEvent.click(screen.getByRole('button', { name: 'Minus 1' }));
            expect(screen.getByTestId('counter')).toHaveTextContent('Index = -1');
        });

        it('Clicking Add 3', () => {
            renderComponent();
            fireEvent.click(screen.getByRole('button', { name: 'Add 3' }));
            expect(screen.getByTestId('counter')).toHaveTextContent('Index = 3');
        });

        it('Clicking Minus 3', () => {
            renderComponent();
            fireEvent.click(screen.getByRole('button', { name: 'Minus 3' }));
            expect(screen.getByTestId('counter')).toHaveTextContent('Index = -3');
        });
    });

    describe('Testing Select Button', () => {
        const fakeContacts = [
            { name: 'Georgie', email: 'Georgie' },
            { name: 'Sheldon', email: 'Sheldon' },
            { name: 'Missy', email: 'Missy' },
        ];

        it('when index is within boundary & index is not changed', () => {
            renderComponent({ testingList: fakeContacts });
            fireEvent.click(
                screen.getByRole('button', { name: 'Select Row based on index above' })
            );
            expect(screen.getByText(/Selected User:/i)).toHaveTextContent('Selected User: Georgie');
        });

        it('when index is within boundary & index is changed', () => {
            renderComponent({ testingList: fakeContacts });
            fireEvent.click(screen.getByRole('button', { name: 'Add 1' }));
            fireEvent.click(
                screen.getByRole('button', { name: 'Select Row based on index above' })
            );
            expect(screen.getByText(/Selected User:/i)).toHaveTextContent('Selected User: Sheldon');
        });

        it('when index is negative', () => {
            renderComponent({ testingList: fakeContacts });
            fireEvent.click(screen.getByRole('button', { name: 'Minus 1' }));
            fireEvent.click(
                screen.getByRole('button', { name: 'Select Row based on index above' })
            );
            expect(screen.getByText(/Selected User:/i)).toHaveTextContent(
                'Selected User: Error!!!!! Index out of bound'
            );
        });

        it('when index is out of boundary', () => {
            renderComponent({ testingList: fakeContacts });
            fireEvent.click(screen.getByRole('button', { name: 'Add 3' }));
            fireEvent.click(
                screen.getByRole('button', { name: 'Select Row based on index above' })
            );
            expect(screen.getByText(/Selected User:/i)).toHaveTextContent(
                'Selected User: Error!!!!! Index out of bound'
            );
        });
    });
});
