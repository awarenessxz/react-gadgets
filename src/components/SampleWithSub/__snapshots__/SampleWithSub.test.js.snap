// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing if component renders properly Snapshot Testing 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Select User from List Below
    </h1>
    <p>
      There are two parts to this Sample: 
    </p>
    <ol>
      <li>
        Click on the Add / Minus buttons to change index value. (to demo testing with multiple same child component in parent component)
      </li>
      <li>
        Select button to select data from a list. (to demo testing data array and error handling)
      </li>
    </ol>
    <h2
      data-testid="counter"
    >
      Index = 0
    </h2>
    <h3>
      Child Component: 
    </h3>
    <button>
      Add 1
    </button>
    <button>
      Minus 1
    </button>
    <h3>
      Child Component: 
    </h3>
    <button>
      Add 3
    </button>
    <button>
      Minus 3
    </button>
    <hr />
    <button>
      Select Row based on index above
    </button>
    <ul>
      <li
        data-testid="usernames"
      >
        Alice
      </li>
      <li
        data-testid="usernames"
      >
        Bruce
      </li>
      <li
        data-testid="usernames"
      >
        Charlie
      </li>
      <li
        data-testid="usernames"
      >
        Donnie
      </li>
      <li
        data-testid="usernames"
      >
        Elizabeth
      </li>
      <li
        data-testid="usernames"
      >
        Frank
      </li>
    </ul>
    <p>
      Selected User: Null
    </p>
  </div>
</DocumentFragment>
`;
